edgelet/aziot-edged/src/workload_manager.rs

```diff
@@ -0,0 +1,287 @@
+    for<'r> &'r <M as ModuleRuntime>::Error: Into<ModuleRuntimeErrorReason>,
+{
+    module_runtime: M,
+    shutdown_senders: HashMap<String, oneshot::Sender<()>>,
+    legacy_workload_uri: Url,
+    home_dir: PathBuf,
+    key_client: Arc<Client>,
+    cert_client: Arc<Mutex<CertificateClient>>,
+    identity_client: Arc<Mutex<IdentityClient>>,
+    config: W,
+}
+
+impl<M, W> WorkloadManager<M, W>
+where
+    W: WorkloadConfig + Clone + Send + Sync + 'static,
+    M: ModuleRuntime + 'static + Authenticator<Request = Request<Body>> + Clone + Send + Sync,
+    <<M as Authenticator>::AuthenticateFuture as Future>::Error: Fail,
+    for<'r> &'r <M as ModuleRuntime>::Error: Into<ModuleRuntimeErrorReason>,
+    <<M as ModuleRuntime>::Module as Module>::Config: Clone + DeserializeOwned + Serialize,
+    <M as ModuleRuntime>::Logs: Into<Body>,
+{
+    pub fn start_manager<F>(
+        settings: &F::Settings,
+        runtime: &M,
+        config: W,
+        tokio_runtime: &mut tokio::runtime::Runtime,
+        create_socket_channel_rcv: UnboundedReceiver<ModuleAction>,
+    ) -> Result<(), Error>
+    where
+        F: MakeModuleRuntime + 'static,
+    {
+        let shutdown_senders: HashMap<String, oneshot::Sender<()>> = HashMap::new();
+
+        let legacy_workload_uri = settings.listen().legacy_workload_uri().clone();
+        let keyd_url = settings.endpoints().aziot_keyd_url().clone();
+        let certd_url = settings.endpoints().aziot_certd_url().clone();
+        let identityd_url = settings.endpoints().aziot_identityd_url().clone();
+
+        let home_dir = settings.homedir().to_path_buf();
+
+        let key_connector = http_common::Connector::new(&keyd_url).expect("Connector");
+        let key_client = Arc::new(aziot_key_client::Client::new(
+            aziot_key_common_http::ApiVersion::V2020_09_01,
+            key_connector,
+        ));
+
+        let cert_client = Arc::new(Mutex::new(cert_client::CertificateClient::new(
+            aziot_cert_common_http::ApiVersion::V2020_09_01,
+            &certd_url,
+        )));
+        let identity_client = Arc::new(Mutex::new(identity_client::IdentityClient::new(
+            aziot_identity_common_http::ApiVersion::V2020_09_01,
+            &identityd_url,
+        )));
+
+        let module_runtime = runtime.clone();
+
+        let workload_manager = WorkloadManager {
+            module_runtime,
+            shutdown_senders,
+            legacy_workload_uri,
+            home_dir,
+            key_client,
+            cert_client,
+            identity_client,
+            config,
+        };
+
+        let module_list: Vec<<M as ModuleRuntime>::Module> =
+            tokio_runtime.block_on(runtime.list()).map_err(|err| {
+                err.context(ErrorKind::Initialize(
+                    InitializeErrorReason::WorkloadService,
+                ))
+            })?;
+
+        tokio_runtime.block_on(futures::future::lazy(move || {
```

Comment:
This is to make it a future. If I remove it, the block on gives me the following error:
the trait bound `[closure@aziot-edged/src/workload_manager.rs:109:32: 112:10 workload_manager:_, module_list:_, create_socket_channel_rcv:_]: futures::Future` is not satisfied
the trait `futures::Future` is not implemented for `[closure@aziot-edged/src/workload_manager.rs:109:32: 112:10 workload_manager:_, module_list:_, create_socket_channel_rcv:_]

Is there a better way?

Reply:
`server()` does its work synchronously, ultimately internally `tokio::spawn()`ing a future. Why does the call to `server` itself need to be a future? Just call it inline.

Reply:
oh, you're right, it is synchronous. Not sure why I though it was not. Thanks for catching that.

Reply:
Actually it is important, there is a good reason for it.
 thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SpawnError { is_shutdown: true }'
 
 Because the server spawn task outside of the tokio context. We discussed it. This was a solution to avoid the above error.